src/Sulu/Component/SmartContent/Orm/DataProviderRepositoryTrait.php

```diff
@@ -295,6 +298,25 @@ protected function append(QueryBuilder $queryBuilder, $alias, $locale, $options
         return [];
     }
 
+    /**
+     * Append select to query builder for "findByFilters" function.
+     *
+     * @param string $alias
+     * @param string $locale
+     * @param array $options
+     *
+     * @return array parameters for query
+     */
+    protected function appendSelect(QueryBuilder $queryBuilder, $alias, $locale, $options = [])
+    {
+        $queryBuilder
+            ->select($alias . '.id')
+            ->distinct()
+            ->orderBy($alias . '.id', 'ASC');
```

Comment:
This does more than just append a select clause... Do we really need to extract that as a separate function?

Reply:
Also, the `createQueryBuilder` function can already be used to something like this, if I am not mistaken. E.g. the [`MediaDataProvider`](https://github.com/sulu/sulu/blob/b48343f/src/Sulu/Bundle/MediaBundle/Entity/MediaDataProviderRepository.php#L205) already makes use of that.

Reply:
But the `->select()` overrides all previous selects 🤔 then we would have to change that to `->addSelect()`

Reply:
The same for the `->orderBy()`

Reply:
The only call I can find of this function is [here](https://github.com/sulu/sulu/pull/5234/files#diff-3ba8d43dcc752376edcd2f5156c711bfR63). So it only seems to be used for initialization. And as mentioned in my previous comment, there is already the `createQueryBuilder` function for that.

Or am I missing something? Is this function also called in other places?

Reply:
At least, in `findByFiltersIds()`, your custom `->select()` or `->orderBy`, that has been set in your custom `DataProviderRepository` in `createQueryBuilder()` would always be overridden [here](https://github.com/sulu/sulu/blob/f5bff6a9cccfc6587b6e59e0a4464ec23b15facf/src/Sulu/Component/SmartContent/Orm/DataProviderRepositoryTrait.php#L62-L64)

Reply:
But if we change `->select()` to `->addSelect()` and `->orderBy()` to `->addOrderBy()`, it wouldn't override previous selects or orderBys

Reply:
I just realized that i create the select depending on `$locale` and `$options`. So i can't use `createQueryBuilder` anyways.

Reply:
Can you link or paste your example code? I simply don't want to have two very similar methods, because it makes things harder to understand. And `createQueryBuilder` is at least sometimes already used in a similar context.

Reply:
https://github.com/sulu/SuluContentBundle/pull/131/files#diff-04d8a1150ad493bbb84cdc1b20be43fcR150

Reply:
Is it just me, or is there really no usage of the `DataProviderRepositoryTrait`?

Actually I am asking because I'd like to see how it is called. Wouldn't it also work if we add `$options` as an argument to the `createQueryBuilder` method?

Reply:
We decided to not use the `DataProviderRepositoryTrait` anymore, so we have full control over the code. Do you still want some of the changes of this pull request to be merged? Then I'm going to adjust it now as you want.

Reply:
May I ask why? Do you need to further optimize your query than this trait would allow you to?

And I think most of the changes in here make sense. I would just integrate them in the `createQueryBuilder` method instead of creating a new one. I think that should be possible, shouldn't it?

Reply:
`createQueryBuilder()` is used by both, `findByFilters()` and `findByFiltersIds()`.

`findByFilters()` selects the whole entity where `id` in `$ids` ordered by `id` `ASC`.

`findByFilterIds()` selects only the distinct `id` without where and also ordered by `id` `ASC`.

The only common code here is the `->orderBy($alias . '.id', 'ASC')`, so we can't extract that code into `createQueryBuilder()` IMO. Also this would be a quite bigger bc break, because it's likely that someone has overridden the `createQueryBuilder()` method.

Reply:
And you want to do that only in the `findByFiltersIds` function? As said, one of the issues here IMO is that it does not only append thee select, but a few things more...

I've had a closer look, and there is already the `append` function in this trait offering the same method signature, and it seems to be called in this function already. Why can't we use that one for what you are trying to implement here?